helix-term/src/compositor.rs

```diff
@@ -136,19 +136,35 @@ impl Compositor {
             keys.push(key.into());
         }
 
+        let mut callbacks = Vec::new();
+        let mut consumed = false;
+
         // propagate events through the layers until we either find a layer that consumes it or we
         // run out of layers (event bubbling)
         for layer in self.layers.iter_mut().rev() {
```

Comment:
since the absence of an inclusive `take_while` in std, rather imperative code here...
could extract a private function `propagate_events` to clean things up a bit:

```rust
    fn propagate_event(&mut self, event: Event, cx: &mut Context) -> (bool, Vec<Callback>) {
        let mut callbacks = Vec::new();

        // propagate events through the layers until we either find a layer that consumes it or we
        // run out of layers (event bubbling)
        for layer in self.layers.iter_mut().rev() {
            match layer.handle_event(event, cx) {
                EventResult::Consumed(Some(callback)) => {
                    callbacks.push(callback);
                    return (true, callbacks);
                }
                EventResult::Consumed(None) => return (true, callbacks),
                EventResult::Ignored(Some(callback)) => {
                    callbacks.push(callback);
                }
                EventResult::Ignored(None) => {}
            };
        }

        (false, callbacks)
    }

    pub fn handle_event(&mut self, event: Event, cx: &mut Context) -> bool {
        // If it is a key event and a macro is being recorded, push the key event to the recording.
        if let (Event::Key(key), Some((_, keys))) = (event, &mut cx.editor.macro_recording) {
            keys.push(key.into());
        }

        let (is_consumed, callbacks) = self.propagate_event(event, cx);
        for callback in callbacks {
            callback(self, cx)
        }

        is_consumed
    }
```

not much of an improvement tho, just keep as is?